File: FindFirst&LastPosInSortedArray/FindFirstAndLastPositionv2.java
public class FindFirstAndLastPositionv2 {
    public int[] searchRange(int[] nums, int target) {
        int first = findBound(nums, target, true);
        if (first == -1) {
            return new int[]{-1, -1};
        }
        int last = findBound(nums, target, false);
        return new int[]{first, last};
    }

    // Helper to find either first or last occurrence
    private int findBound(int[] nums, int target, boolean findFirst) {
        int low = 0, high = nums.length - 1, result = -1;

        while (low <= high) {
            int mid = low + (high - low) / 2;

            if (nums[mid] == target) {
                result = mid;
                if (findFirst) {
                    high = mid - 1; // keep searching to the left
                } else {
                    low = mid + 1; // keep searching to the right
                }
            } else if (nums[mid] < target) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }

        return result;
    }
}

File: 34. FindFirst&LastPosInSortedArray/FindFirstAndLastPositionv2.py
def FindFirstAndLastPosition(nums, target):
    def find_left(nums, target):
        left, right = 0, len(nums) - 1
        first = -1
        while left <= right:
            mid = (left + right) // 2
            if nums[mid] == target:
                first = mid
                right = mid - 1  # sigue buscando más a la izquierda
            elif nums[mid] < target:
                left = mid + 1
            else:
                right = mid - 1
        return first

    def find_right(nums, target):
        left, right = 0, len(nums) - 1
        last = -1
        while left <= right:
            mid = (left + right) // 2
            if nums[mid] == target:
                last = mid
                left = mid + 1  # sigue buscando más a la derecha
            elif nums[mid] < target:
                left = mid + 1
            else:
                right = mid - 1
        return last

    return [find_left(nums, target), find_right(nums, target)]


if __name__ == "__main__":
    nums = list(map(int, input("Enter the sorted array (space-separated): ").split()))
    target = int(input("Enter the target value: "))
    result = FindFirstAndLastPosition(nums, target)
    print(f"First and Last Position of {target} in the array: {result}")
